docs/design/cloud-init.rst

```diff
@@ -0,0 +1,260 @@
+
+> Vendordata is data provided by the entity that launches an instance (for
+> example, the cloud provider). This data can be used to customize the image to
+> fit into the particular environment it is being run in.
+
+Unfortunately, older versions of cloud-init don't play well with
+``vendor-data``, which leaves us with ``meta-data`` and the files under
+``/etc/cloud/cloud.cfg.d``.
+
+Configuration Tasks
+-------------------
+
+*snf-image* configures the VM by running a number of configuration tasks on the
+hard disk of the VM. On images that support cloud-init, instead of directly
+altering the needed system files, *snf-image-helper* could add cloud-init
+configuration into the meta-data file as well as files under
+``/etc/cloud/cloud.cdg.d/``.
+
+The proposed changes for the configuration tasks are:
+
+**10FixPartitionTable**: The task should be prevented from running. Cloud-init
+will automatically grow the partition to consume the available space using the
+`growroot` module of cloud-initramfs-tools and `cc_growpart` module.
+
+**20FilesystemResizeUnmounted**: This task should be prevented from running.
+Cloud-init will automatically enlarge the file system using the `cc_resizefs`
+module.
+ 
+**30MountImage**: No changes are needed for this task.
+
+**35InstallUnattend**: This can stay intact of be prevented from running. It's
+Windows specific task and won't do anything against a Linux image.
+
+**40FilesystemResizeMounted**: This should be prevented from running.
+Cloud-init will use ``cc_resizefs`` module to grow the file system.
+
+**50AddSwap**: We can use this to create a configuration file to add swap,
+using the ``cc_mount`` module. This module has a swap config key that can be
+used like this:
+```
+swap:
+    filename: <file>
+    size: <"auto"/size in bytes>
+    maxsize: <size in bytes>
+```
+With this key we can add a swap file. The SWAP image property either expects a
+<partition id>:<size> tuple as value or a letter denoting that a whole disk
+should be used as swap. In cloud-init, since we don't do the partitioning
+ourselves, we cannot create the swap partition. In this case we can use the
+"swap" key to create a swap file and ignore the partition id. On the other
+hand, if we get a request to make a disk as swap, we should be able to use the
+``mounts`` key of the ``cc_mounts`` module:
+```
+mounts:
+    - [ /dev/ephemeral0, /mnt, auto, "defaults,noexec" ]
+    - [ sdc, /opt/data ]
+    - [ xvdh, /opt/data, "auto", "defaults,nofail", "0", "0" ]
+```
+With this module we could create the swap entry for the disk. The problem is
+that we are using non ephemeral device names (UUID for swap) and this module
+does not support this. In order to know the correct device naming we should
+forward the ``DISK_%N_FRONTEND_TYPE`` from *Ganeti* to *snf-image-helper* which
+we don't. For now, its better if we bypass cloud-init and put the swap entry
+directly to fstab, if the swap is a whole disk.
+
```

Comment:
Is it decided that there will be swap files instead of swap partitions for cloud-init based images ? If yes, the above text needs rephrasing to clarify that.